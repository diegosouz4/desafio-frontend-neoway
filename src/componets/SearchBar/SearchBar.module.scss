.form {
  background-color: var(--color-white-100);
  border-radius: 0.25rem;
  box-shadow: 0 0.5rem 1.25rem -0.25rem rgba(0, 0, 0, 0.25);
  width: 90%;
  max-width: 800px;
  margin: 0 auto;
  padding: 0.625rem;
  position: relative;
  transition: 450ms var(--pop);

  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: space-between;
  gap: 0.3125rem;

  &:hover,
  &:focus {
    transform: scale(1.01);
  }

  fieldset {
    border: 0;
    padding: 0.3125rem;
    position: relative;
    &:nth-child(1) {
      flex-grow: 1;
    }

    @media (max-width: 680px) {
      flex-grow: 1;
      &:nth-child(1) {
        width: 100%;
      }
    }

    label {
      color: var(--color-text);
      font-size: 0.75rem;
      line-height: 1.1;
      position: absolute;
      top: -0.3125rem;
      left: 0.625rem;
    }

    .input,
    .select {
      color: var(--color-text);
      border: 0;
      border-radius: 0.25rem;
      border: 2px solid transparent;
      background-color: #cbcbcb;
      height: 2.5rem;
      padding: 0.3125rem 0.625rem;
      width: 100%;

      &:focus,
      &:hover {
        border: 2px solid var(--color-main);
        outline-color: var(--color-main);
      }
    }
  }

  .btn {
    display: block;
    color: var(--color-main-200);
    background-color: var(--color-main);
    font-size: 0.875rem;
    font-weight: 600;
    border-radius: 0.25rem;
    padding: 0.75rem 1.25rem;
    transition: 450ms var(--pop);

    position: relative;
    &:hover,
    &:focus {
      transform: scale(1.04);
      color: var(--color-main);
      background-color: var(--color-main-200);
    }
  }
}
