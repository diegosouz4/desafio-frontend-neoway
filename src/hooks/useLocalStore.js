export function SetLocalStore(key, state) {
  if (key !== undefined) {
    window.localStorage.setItem(key, JSON.stringify(state));
    return state;
  }

  return null;
}

export function GetLocalStore(key) {
  const valueLocalStore = window.localStorage.getItem(key);
  if (!key || !valueLocalStore) return null;

  return JSON.parse(valueLocalStore);
}
