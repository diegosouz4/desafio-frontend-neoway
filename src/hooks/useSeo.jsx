export default function useSeo(title, description) {
  if (title !== undefined && title !== null) {
    document.title = title;
  }

  if (description !== undefined && description !== null) {
    document.querySelector("meta[name='description']").content = description;
  }

  return [title, description];
}
