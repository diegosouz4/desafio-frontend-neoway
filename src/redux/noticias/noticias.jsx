import { createSlice } from "@reduxjs/toolkit";

const initialState = {
  news: {
    loading: false,
    data: null,
    error: null,
  },
};

const slice = createSlice({
  name: "noticias",
  initialState: initialState,
  reducers: {
    fetchNewsStart(state) {
      (state.news.loading = true),
        (state.news.data = null),
        (state.news.error = null);
    },
    fetchNewsSuccess(state, action) {
      (state.news.loading = false),
        (state.news.data = action.payload),
        (state.news.error = null);
    },
    fetchNewsError(state, action) {
      (state.news.loading = false),
        (state.news.data = null),
        (state.news.error = action.payload);
    },
  },
});

const { fetchNewsStart, fetchNewsSuccess, fetchNewsError } = slice.actions;

export const fetchNoticias = (query) => async (dispatch) => {
  const { q, language, sortBy } = query;
  const url = `https://newsapi.org/v2/everything?q=${q}&language=${language}&sortBy=${sortBy}&from=2023-09-24&apiKey=${
    import.meta.env.VITE_API_TOKEN_TESTE
  }&pageSize=100`;

  try {
    dispatch(fetchNewsStart());
    const response = await fetch(url);
    const data = await response.json();
    //Removendo as notícias
    data.articles = await data.articles.filter(
      (item) => item.title !== "[Removed]" && item.title !== null
    );
    //Alterando os id's null pelo index da notícia.
    data.articles = await data.articles.map((item, index) =>({...item, source: {...item.source, id: index}}));
    return dispatch(fetchNewsSuccess(data));
  } catch (error) {
    dispatch(fetchNewsError(error.message));
  }
};

export default slice.reducer;
