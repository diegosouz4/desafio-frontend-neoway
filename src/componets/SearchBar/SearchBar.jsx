import React from "react";
import { useDispatch } from "react-redux";
import { fetchNoticias } from "../../redux/noticias/noticias";
import style from "./SearchBar.module.scss";
import { SetLocalStore } from "../../hooks/useLocalStore";

const initialState = {
  q: "",
  language: "pt",
  sortBy: "relevancy",
};

export default function SearchBar() {
  const [query, setQuery] = React.useState(initialState);
  const dispatch = useDispatch();

  function handleSubmit(e) {
    e.preventDefault();
    dispatch(fetchNoticias(query));
    SetLocalStore("query", query);
  }

  return (
    <form data-testid="form" className={style.form} onSubmit={handleSubmit}>
      <fieldset>
        <label htmlFor="q">Buscar</label>
        <input
          className={style.input}
          name="q"
          type="text"
          placeholder="buscar..."
          value={query.q}
          onChange={({ target }) =>
            setQuery((prev) => ({ ...prev, q: target.value }))
          }
          required
        />
      </fieldset>

      <fieldset>
        <label htmlFor="language">Qual o idioma?</label>
        <select
          className={style.select}
          id="language"
          value={query.language}
          onChange={({ target }) =>
            setQuery((prev) => {
              return { ...prev, language: target.value };
            })
          }
        >
          <option value="en">EN</option>
          <option value="es">ES</option>
          <option value="fr">FR</option>
          <option value="pt">PT</option>
        </select>
      </fieldset>

      <fieldset>
        <label htmlFor="filterBy">Filtrar por:</label>
        <select
          id="filterBy"
          className={style.select}
          value={query.sortBy}
          onChange={({ target }) =>
            setQuery((prev) => {
              return { ...prev, sortBy: target.value };
            })
          }
        >
          <option value="relevancy">Relevância</option>
          <option value="popularity">Popularidade</option>
          <option value="publishedAt">Data da publicação</option>
        </select>
      </fieldset>

      <button className={style.btn}>Buscar</button>
    </form>
  );
}
