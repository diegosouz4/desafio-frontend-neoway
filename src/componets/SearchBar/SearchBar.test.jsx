import { fireEvent, render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { useDispatch, Provider } from "react-redux";
import configureStore from "redux-mock-store";
import SearchBar from "./SearchBar";
import { fetchNoticias } from "../../redux/noticias/noticias";
import { SetLocalStore } from "../../hooks/useLocalStore";

jest.mock("../../redux/noticias/noticias", () => {
  return { fetchNoticias: jest.fn() };
});

jest.mock("../../hooks/useLocalStore", () => {
  return { SetLocalStore: jest.fn() };
});

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

const mockStore = configureStore([]);

describe("<SearchBar />", () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      noticias: {
        news: {
          data: {
            articles: [],
          },
          likes: [],
        },
      },
    });
  });

  it("should SearchBar was rendered", () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Provider store={store}>
          <SearchBar />
        </Provider>
      </BrowserRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("test submit", async () => {
    const dispatch = jest.fn();
    fetchNoticias.mockReturnValue(dispatch);
    useDispatch.mockReturnValue(dispatch);

    render(
      <BrowserRouter>
        <Provider store={store}>
          <SearchBar />
        </Provider>
      </BrowserRouter>
    );

    const formEl = screen.getByTestId("form");

    const inputQuery = screen.getByPlaceholderText("buscar...");
    const inputLang = screen.getByLabelText("Qual o idioma?");
    const inputFilterBy = screen.getByLabelText("Filtrar por:");

    await fireEvent.change(inputQuery, { target: { value: "Jest" } });
    await fireEvent.change(inputLang, { target: { value: "en" } });
    await fireEvent.change(inputFilterBy, { target: { value: "publishedAt" } });

    await fireEvent.submit(formEl);

    expect(fetchNoticias).toHaveBeenCalledWith({
      q: "Jest",
      language: "en",
      sortBy: "publishedAt",
    });

    expect(SetLocalStore).toHaveBeenCalledWith("query", {
      q: "Jest",
      language: "en",
      sortBy: "publishedAt",
    });
  });
});
