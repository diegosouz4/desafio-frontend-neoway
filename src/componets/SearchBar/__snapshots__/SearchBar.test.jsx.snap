// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> should SearchBar was rendered 1`] = `
<DocumentFragment>
  <form
    data-testid="form"
  >
    <fieldset>
      <label
        for="q"
      >
        Buscar
      </label>
      <input
        name="q"
        placeholder="buscar..."
        required=""
        type="text"
        value=""
      />
    </fieldset>
    <fieldset>
      <label
        for="language"
      >
        Qual o idioma?
      </label>
      <select
        id="language"
      >
        <option
          value="en"
        >
          EN
        </option>
        <option
          value="es"
        >
          ES
        </option>
        <option
          value="fr"
        >
          FR
        </option>
        <option
          value="pt"
        >
          PT
        </option>
      </select>
    </fieldset>
    <fieldset>
      <label
        for="filterBy"
      >
        Filtrar por:
      </label>
      <select
        id="filterBy"
      >
        <option
          value="relevancy"
        >
          Relevância
        </option>
        <option
          value="popularity"
        >
          Popularidade
        </option>
        <option
          value="publishedAt"
        >
          Data da publicação
        </option>
      </select>
    </fieldset>
    <button>
      Buscar
    </button>
  </form>
</DocumentFragment>
`;
